drivers/i2c/i2c_mcux_flexcomm.c

```diff
@@ -345,18 +404,39 @@ int mcux_flexcomm_target_unregister(const struct device *dev,
 	struct mcux_flexcomm_data *data = dev->data;
 	struct mcux_flexcomm_target_data *target;
 	I2C_Type *base = config->base;
+	uint32_t clock_freq;
+
+	/* Get the clock frequency */
+	if (clock_control_get_rate(config->clock_dev, config->clock_subsys,
+				   &clock_freq)) {
+		return -EINVAL;
+	}
 
 	target = mcux_flexcomm_find_target_by_address(data, target_config->address);
 	if (!target || !target->target_attached) {
 		return -EINVAL;
 	}
 
+	if (mcux_flexcomm_setup_i2c_config_address(data, target, 0x00, true) < 0) {
+		return -EINVAL;
+	}
+
 	target->target_cfg = NULL;
 	target->target_attached = false;
 
 	data->nr_targets_attached--;
 
-	I2C_SlaveDeinit(base);
+	if (data->nr_targets_attached == 0) {
+		I2C_SlaveDeinit(base);
+	} else {
+		/* still slaves attached, reconfigure the I2C peripheral after address removal */
+		I2C_SlaveInit(base, &data->i2c_cfg, clock_freq);
+		I2C_SlaveTransferCreateHandle(base, &data->target_handle,
+				i2c_target_transfer_callback, data);
+		I2C_SlaveTransferNonBlocking(base, &data->target_handle,
+				kI2C_SlaveCompletionEvent | kI2C_SlaveTransmitEvent |
+				kI2C_SlaveReceiveEvent | kI2C_SlaveDeselectedEvent);
```

Comment:
NIT: this logic could be moved to a helper function so it can be called both from here and from mcux_flexcomm_target_register() to ensure they stay in sync